File: 253-meeting-rooms-ii/meeting-rooms-ii.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0

        intervals.sort(key=lambda x:x[0])
        min_heap = []

        for interval in intervals:
            start, end = interval

            if min_heap and min_heap[0] <= start:
                heappop(min_heap)

            heappush(min_heap, end)

        return len(min_heap)

File: 253-meeting-rooms-ii/meeting-rooms-ii.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {

        if (intervals.length == 0 || intervals.equals(null)) {
            return 0;
        }

        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        PriorityQueue<Integer> min_heap = new PriorityQueue<>();

        for (int[] interval : intervals){
            int start = interval[0];
            int end = interval[1];

            if (min_heap.size() > 0 && min_heap.peek() <= start) {
                min_heap.poll();
            }

            min_heap.add(end);
        }

        return min_heap.size();

        // for (int[] interval : intervals){
        //     System.out.println(interval[0]);

        // }

        // return 0;
    }
}